src_python/habitat_sim/simulator.py

```diff
@@ -423,27 +680,32 @@ def get_sensor_observations(
 
     def get_sensor_observations(
         self, agent_ids: Union[int, List[int]] = 0
-    ) -> Union[ObservationDict, Dict[int, ObservationDict],]:
+    ) -> Union[ObservationDict, Dict[int, ObservationDict]]:
+
         if isinstance(agent_ids, int):
             agent_ids = [agent_ids]
             return_single = True
         else:
             return_single = False
 
         for agent_id in agent_ids:
-            agent_sensorsuite = self.__sensors[agent_id]
-            for _sensor_uuid, sensor in agent_sensorsuite.items():
-                sensor.draw_observation()
+            agent_sensors = self.get_agent(agent_id).scene_node.subtree_sensors
+            for _, sensor in agent_sensors.items():
+                sensor.draw_observation(self)
 
         # As backport. All Dicts are ordered in Python >= 3.7
         observations: Dict[int, ObservationDict] = OrderedDict()
-        for agent_id in agent_ids:
-            agent_observations: ObservationDict = {}
-            for sensor_uuid, sensor in self.__sensors[agent_id].items():
-                agent_observations[sensor_uuid] = sensor.get_observation()
-            observations[agent_id] = agent_observations
+        if agent_id >= 0 and agent_id < len(self.__sensors):
+            for agent_id in agent_ids:
+                agent_observations: ObservationDict = {}
+                for sensor_uuid, sensor in self.__sensors[agent_id].items():
+                    agent_observations[sensor_uuid] = sensor.get_observation()
```

Comment:
This should probably turned into a dict comprehension as that will be a bit faster.